apps/system/flash_test/flash_test.c

```diff
@@ -53,44 +53,105 @@
 /****************************************************************************
  * Included Files
  ****************************************************************************/
-
-#include <tinyara/config.h>
-#include <tinyara/progmem.h>
-#include <sys/stat.h>
-
-#include <unistd.h>
 #include <stdio.h>
+#include <string.h>
 #include <stdlib.h>
+#include <errno.h>
+
 #include <tinyara/fs/mtd.h>
 
 /****************************************************************************
- * Pre-processor Definitions
+ * Definitions
  ****************************************************************************/
+#define BUF_SIZE   1024
 
 /****************************************************************************
- * Private data
+ * Private Data
  ****************************************************************************/
 
 /****************************************************************************
- * Public Functions
+ * Private Functions
  ****************************************************************************/
+static int flash_read(unsigned long addr, int size)
+{
+	int ret;
+	int offset;
+	int buf_idx;
+	int read_size;
+	int remaining;
+	FAR uint8_t *buffer;
+	FAR struct mtd_dev_s *dev_mtd = NULL;
+
+	if (addr == 0 || size <= 0) {
+		printf("Invalid arguments, offset 0x%x, read size %d\n", addr, size);
+		return ERROR;
+	}
+
+	dev_mtd = up_flashinitialize();
+	if (!dev_mtd) {
+		printf("up_flashinitialize Failed\n");
+		return ERROR;
+	}
+
+	buffer = (uint8_t *)malloc(BUF_SIZE);
+	if (!buffer) {
+		printf("Failed to allocate buffer %d\n", BUF_SIZE);
+		return ERROR;
+	}
+
+	offset = 0;
+	remaining = size;
 
+	printf("Read 0x%x: ", addr);
+
+	while (remaining > 0) {
+		read_size = BUF_SIZE < remaining ? BUF_SIZE : remaining;
+		ret = MTD_READ(dev_mtd, addr + offset, read_size, buffer);
+		if (ret <= 0) {
+			printf("Read Failed : %d\n", ret);
+			free(buffer);
+			return ERROR;
+		}
+
+		/* Print read buffer */
+		for (buf_idx = 0; buf_idx < ret; buf_idx++) {
+			if ((buf_idx % 16) == 0) {
+				printf("\n");
+			}
+			printf("0x%02x ", buffer[buf_idx]);
+		}
+		offset += ret;
+		remaining -= ret;
+	}
+	printf("\n");
+
+	free(buffer);
+
+	return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
 #ifdef CONFIG_BUILD_KERNEL
 int main(int argc, FAR char *argv[])
 #else
-int flash_eraseall_main(int argc, char *argv[])
+int flash_test_main(int argc, char *argv[])
 #endif
 {
-	/* Argument given? */
+	off_t offset;
+	int read_size;
 
-	if (argc < 2) {
-		fprintf(stderr, "usage: flash_eraseall flash_block_device\n");
-		return -1;
-	}
+	if (argc == 4 && !strncmp(argv[1], "read", strlen("read") + 1)) {
```

Comment:
If you support only read option, is checking argv[1] needed?
Or do you consider expansion?

Reply:
Yes. I think some tests for flash can be added here.